### PR TITLE
Lesson 2 linked list

### LINKED ISSUE #8: Lesson 2 Home task 1, linked list reverse

### PR DESCRIPTION
**Pull request by AndriiWell**
Linked list via iterations and recursivelly.

### CODE REVIEW

File: src/main/java/msjava2lesson/Main.java [hunk 1]
```diff
@@ -1,7 +1,370 @@
...
 public class Main {
+
     public static void main(String[] args) {
-        System.out.println("Hello World!");
+
+        taskReverseLinkedList();
+        taskMoveZeros();
+    }
+
+    private static void linkedListPrint(ListNode nodeStart) {
```
Comment (b-chupika) [inline src/main/java/msjava2lesson/Main.java]:
```suggestion
    private static void linkedListPrint(ListNode head) {
```

File: src/main/java/msjava2lesson/Main.java [hunk 2]
```diff
@@ -1,7 +1,370 @@
...
+
+        boolean isStart = true;
+        while (nodeStart != null){
+
+            if(isStart)
+                System.out.print("Head:");
```
Comment (b-chupika) [inline src/main/java/msjava2lesson/Main.java]: Could it be moved out of the while loop?

File: src/main/java/msjava2lesson/Main.java [hunk 3]
```diff
@@ -1,7 +1,370 @@
...
+    private static void zeroElementInLinkedListAndReverse() {
+        ListNode nodeOneElement = null;
+
+        // Show original list.
+        System.out.println("iteratively");
+        System.out.print("Origin:   ");
+        linkedListPrint(nodeOneElement);
+
+        // Reverse
+        msjava2lesson.reverselinkedlist.Solution reverseSolution = new msjava2lesson.reverselinkedlist.Solution();
```
Comment (b-chupika) [inline src/main/java/msjava2lesson/Main.java:53]: No need to specify the full path. If you have two classes with name `Solution` consider renaming them
```suggestion
        Solution reverseSolution = new Solution();
```

File: src/main/java/msjava2lesson/moveZeroes/Solution.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+        for (int iNum = 0; iNum < iNumLength; iNum++) {
+            iIterateValue = aNum[iNum];
+            if(iValueToMoveToRear == iIterateValue) {
+                if(iEarliestValueIndexToMove == -1)
+                    iEarliestValueIndexToMove = iNum;
+            }
+            else {
+                if(iEarliestValueIndexToMove == -1) // case when array starts not from the iValueToMoveToRear value.
+                   continue;
```
Comment (b-chupika) [inline src/main/java/msjava2lesson/moveZeroes/Solution.java:27]:
```suggestion
                if(iEarliestValueIndexToMove == -1) { // case when array starts not from the iValueToMoveToRear value.
                   continue;
                }   
```

File: src/main/java/msjava2lesson/moveZeroes/Solution.java [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+
+        int iValueToMoveToRear = 0; // Here may be any value, the zero value is just the task condition.
+        int iEarliestValueIndexToMove = -1; // Index of value to move to rear of array.
+        int iIterateValue; // Current value in the cycle iteration.
+
+        for (int iNum = 0; iNum < iNumLength; iNum++) {
+            iIterateValue = aNum[iNum];
+            if(iValueToMoveToRear == iIterateValue) {
+                if(iEarliestValueIndexToMove == -1)
+                    iEarliestValueIndexToMove = iNum;
```
Comment (b-chupika) [inline src/main/java/msjava2lesson/moveZeroes/Solution.java:23]: Please use curly braces. https://google.github.io/styleguide/javaguide.html#s4.1.1-braces-always-used
```suggestion
                if(iEarliestValueIndexToMove == -1) {
                    iEarliestValueIndexToMove = iNum;
                }
```

File: src/main/java/msjava2lesson/reverselinkedlist/ListNode.java [hunk 6]
```diff
@@ -0,0 +1,19 @@
+package msjava2lesson.reverselinkedlist;
+
+public class ListNode{
+
+    public int val;
+
+    public ListNode next;
```
Comment (b-chupika) [inline src/main/java/msjava2lesson/reverselinkedlist/ListNode.java:7]: by default, in Java we we have `private` access modifier for fields
```suggestion
    private int val;

    private ListNode next;
```

File: src/main/java/msjava2lesson/reverselinkedlist/ListNode.java [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+
+    ListNode(){}
```
Comment (b-chupika) [inline src/main/java/msjava2lesson/reverselinkedlist/ListNode.java]: why no access modifier here?

**Review by b-chupika — APPROVED** [review]
Good job!

File: src/main/java/msjava2lesson/Main.java (hunk 3 above)
Comment (MukBet) [reply src/main/java/msjava2lesson/Main.java:56]: I have class named Solution in two namespaces: 
 - reverselinkedlist
 - moveZeroes
So,I am not sure I can do like you suggested.

Comment (MukBet) [reply src/main/java/msjava2lesson/Main.java:56]: I have class named Solution in two namespaces:

reverselinkedlist
moveZeroes
So,I am not sure I can do like you suggested.

Comment (AndriiWell) [reply src/main/java/msjava2lesson/Main.java:53]: I have the same class named Solution in two different namespaces:
 - moveZeroes
 - reverselinkedlist
 So, I think it is incorrect to do like you suggested in the Main class.

File: src/main/java/msjava2lesson/reverselinkedlist/ListNode.java (hunk 7 above)
Comment (AndriiWell) [reply src/main/java/msjava2lesson/reverselinkedlist/ListNode.java]: It is default structure! I do not use this method at all. Removed.

File: src/main/java/msjava2lesson/reverselinkedlist/ListNode.java (hunk 6 above)
Comment (AndriiWell) [reply src/main/java/msjava2lesson/reverselinkedlist/ListNode.java:7]: All fields are public for that class - I use this fields in code of Main class without setters.

File: src/main/java/msjava2lesson/moveZeroes/Solution.java (hunk 5 above)
Comment (MukBet) [reply src/main/java/msjava2lesson/moveZeroes/Solution.java:23]: Fixed

**Pull request merged**